### PR TITLE
Fix cycle dispute-coordinator <-> dispute-distribution 

### PR DESCRIPTION
**Pull request by eskimor**
Refactored, so dispute-distribution needs to communicate less back. The remaining communication is now happening in a background task and limited to one at a time.

I also refactored dispute-coordinator a bit, so initialized and non-initialized state use the same code and cannot so easily get out of sync anymore (it has been, which is also fixed in this PR).

- [x] Fix dispute-distribution tests
- [x] Update guide
- [x] Possibly more whitespace fixes

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2209402

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2209466

File: node/network/dispute-distribution/src/sender/mod.rs [hunk 1]
```diff
@@ -122,7 +142,7 @@ impl DisputeSender {
 					ctx,
 					runtime,
 					&self.active_sessions,
-					self.tx.clone(),
+					NestingSender::new(self.tx.clone(), DisputeSenderMessage::TaskFinish),
```
Comment (tdimitrov) [inline node/network/dispute-distribution/src/sender/mod.rs:145]: I'm missing something here. `NestingSender::new` has got `(parent: NestingSender<M, M2>, child_conversion: fn(M1) -> M2)` as input parameters. 

How `DisputeSenderMessage::TaskFinish` can be `fn(M1) -> M2`?

File: node/subsystem-util/src/nesting_sender.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Hierarchical messages.
```
Comment (mrcnski) [inline node/subsystem-util/src/nesting_sender.rs:17]: First off, these docs are excellent! 😃

nit: Some section headings would be handy here for more structure. Maybe `# Background`, `# Architecture`, and `# Limitations`.

File: node/subsystem-util/src/nesting_sender.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+//! to spawn and receive messages, which in turn do the same, still the root logic stays unchanged.
+//! Everything is isolated to the level where it belongs, while we still keep a single task scope
+//! in all non blocking/not CPU intensive parts, which allows us to share data via references for
+//! example.
+//!
+//! Because the wrapping is optional and transparent to the lower modules, each module can also be
+//! used at the top directly without any wrapping, e.g. for standalone use or for testing purposes.
+//!
+//! In the interest of time I refrain from providing a usage example here for now, but would instead
+//! like to point you to the dispute-distribution subsystem which makes use of this architecture.
```
Comment (mrcnski) [inline node/subsystem-util/src/nesting_sender.rs:67]: nit: I was actually thinking before I got this to this part that an example would be great. 😅 I would like to see one here eventually. We should have a TODO or an issue for later.

File: node/subsystem-util/src/nesting_sender.rs [hunk 4]
```diff
@@ -0,0 +1,148 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
```
Comment (mrcnski) [inline node/subsystem-util/src/nesting_sender.rs:1]: nit: I'm going to be that guy:

```suggestion
// Copyright 2023 Parity Technologies (UK) Ltd.
```

File: node/subsystem-util/src/nesting_sender.rs [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+//! for concurrency and parallelism when there is an actual need, not just because we can and it is
+//! reasonably safe in Rust.
+//!
+//! I very much agree with many points in this blog post for example:
+//!
+//! https://maciej.codes/2022-06-09-local-async.html
+//!
+//! Another very good post by Pierre (Tomaka):
+//!
+//! https://tomaka.medium.com/a-look-back-at-asynchronous-rust-d54d63934a1c
```
Comment (mrcnski) [inline node/subsystem-util/src/nesting_sender.rs:30]: nit: I noticed that these don't get turned into links:

```suggestion
//! <https://maciej.codes/2022-06-09-local-async.html>
//!
//! Another very good post by Pierre (Tomaka):
//!
//! <https://tomaka.medium.com/a-look-back-at-asynchronous-rust-d54d63934a1c>
```

File: node/subsystem-util/src/lib.rs [hunk 6]
```diff
@@ -72,6 +72,12 @@ pub mod runtime;
 /// Database trait for subsystem.
 pub mod database;
 
+/// Nested message sending.
+///
+/// Useful for having mostly synchronous code, with submodules spawning short lived asynchronous
+/// tasks, sending messages back.
+pub mod nesting_sender;
```
Comment (mrcnski) [inline node/subsystem-util/src/lib.rs:79]: nit: I noticed in the generated rustdoc that everything after the first line is not included. I think we need something like this (unformatted):

```suggestion
/// Nested message sending. Useful for having mostly synchronous code, with submodules spawning short lived asynchronous
/// tasks, sending messages back.
pub mod nesting_sender;
```

File: node/subsystem-util/src/nesting_sender.rs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+//! With the types and traits in this module you can achieve exactly that: You write modules which
+//! just execute logic and can call into the functions of other modules - yes we are calling normal
+//! functions. For the case a module you are calling into requires an occasional background task,
+//! you provide it with a `NestingSender<M, ChildModuleMessage>` that it can pass to any spawned tasks.
+//!
+//! This way you don't have to spawn a task for each module just for it to be able to handle
+//! asynchronous events. The module relies on the using/enclosing code/module to forward it any
+//! asynchronous messages in a structured way.
+//!
+//! What makes this architecture nice is the separation of concerns - at the top you only have to
```
Comment (mrcnski) [inline node/subsystem-util/src/nesting_sender.rs:55]: nit: Just some quick feedback here, and maybe it's just me. But not having looked at any code or examples yet, I was a little confused where the nesting comes in. I think I get it after a couple re-reads, but could be more clear.

File: node/subsystem-util/src/nesting_sender.rs [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+{
+	/// Create a new `NestingSender` which wraps a given "parent" sender.
+	///
+	/// Doing the necessary conversion from the child message type `M1` to the parent message type
+	/// `M2.
+	/// M1 -> M2 -> M
+	/// Inputs:
+	///		F(M2) -> M (via parent)
+	///		F(M1) -> M2 (via child_conversion)
+	/// Result: F(M1) -> M
```
Comment (mrcnski) [inline node/subsystem-util/src/nesting_sender.rs:119]: The formatting here is a bit messed up. I tried to fix it.

```suggestion
	/// This takes care of the necessary conversion from the child message type `M1` to the parent
	/// message type `M2`.
	///
	/// ```text
	/// M1 -> M2 -> M
	/// Inputs:
	///		F(M2) -> M (via parent)
	///		F(M1) -> M2 (via child_conversion)
	/// Result: F(M1) -> M
	/// ```
```

File: roadmap/implementers-guide/src/node/disputes/dispute-coordinator.md [hunk 9]
```diff
@@ -681,12 +681,17 @@ struct State {
...
-When the subsystem is initialised it waits for a new leaf (message `OverseerSignal::ActiveLeaves`).
-The leaf is used to initialise a `RollingSessionWindow` instance (contains leaf hash and
-`DISPUTE_WINDOW` which is a constant.
 
-Next the active disputes are loaded from the DB. The subsystem checks if there are disputes for
-which a local statement is not issued. A list of these is passed to the main loop.
+When the subsystem is initialised it waits for a new leaf (message
+`OverseerSignal::ActiveLeaves`). The leaf is used to initialise a
+`RollingSessionWindow` instance (contains leaf hash and `DISPUTE_WINDOW` which
+is a constant.
```
Comment (mrcnski) [inline roadmap/implementers-guide/src/node/disputes/dispute-coordinator.md:688]: nit:

```suggestion
`RollingSessionWindow` instance (contains leaf hash and `DISPUTE_WINDOW` which
is a constant).
```

File: node/core/dispute-coordinator/src/import.rs [hunk 10]
```diff
@@ -347,6 +365,11 @@ impl<V> CandidateVoteState<V> {
 		self.dispute_status.map_or(false, |s| s.has_concluded_against())
 	}
 
+	/// Has concluded. Either for or agains the candidate.
+	pub fn has_concluded(&self) -> bool {
+		self.has_concluded_against() || self.has_concluded_for()
+	}
```
Comment (mrcnski) [inline node/core/dispute-coordinator/src/import.rs:371]: Is this used anywhere?

**Review by mrcnski — APPROVED** [review]
It's my first time looking at these subsystems, and I'm not great at protocols yet, but I think it looks good!

File: node/subsystem-util/src/nesting_sender.rs [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+		M2: 'static,
+	{
+		let NestingSender { sender, conversion } = parent;
+		Self { sender, conversion: Arc::new(move |x| conversion(child_conversion(x))) }
+	}
+
+	/// Send a message via the underlying mpsc.
+	///
+	/// Necessary conversion is accomplished.
+	pub async fn send_message(&mut self, m: M1) -> Result<(), mpsc::SendError> {
```
[the remark this answers is not in the document]
Comment (eskimor) [reply node/subsystem-util/src/nesting_sender.rs:131]: I feel `feed` is just an artifact of (ab)using a very generic trait `Sink` for the channel interface. `send` does well a send and a flush. `feed` does not flush. This is intended for disk or network IO, where you want to batch multiple writes together into a single flush. For channels, this does not make any sense. `flush` just waits for the channel to have capacity after the feed. Which is pointless ... this effectively only reduces channel capacity by one.

Hence, I just called it more verbosely `send_message` to (hopefully) avoid confusion.

File: node/subsystem-util/src/lib.rs (hunk 6 above)
Comment (eskimor) [reply node/subsystem-util/src/lib.rs:79]: it is, but not in the overview. Fixed.

File: node/subsystem-util/src/nesting_sender.rs (hunk 7 above)
Comment (eskimor) [reply node/subsystem-util/src/nesting_sender.rs:57]: Added example, hope that helps.

File: node/subsystem-util/src/nesting_sender.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (tdimitrov) [reply node/subsystem-util/src/nesting_sender.rs]: I see. Thanks!

**Review by tdimitrov — APPROVED** [review]
I learnt some new stuff reviewing this PR. I'm also not knowledgeable enough to do a meaningful code review. It looks good from my point of view so I'm approving.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2242824

**Pull request merged**